File: dotnet/src/xjustiz.core-dotnet/Util/Versioning/CompatibilityChecker.cs
namespace xjustiz.core_dotnet.Util.Versioning;

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using xjustiz.core_dotnet.Models;

/// <summary>
/// Überprüft die Kompatibilität von X.Justiz-Nachrichten mit verschiedenen X.Justiz- und X.Justiz-Core-Versionen.<br/>
/// <u><b>Compatibility Checker:</b></u> Checks the compatibility of X.Justiz messages with different X.Justiz and X.Justiz Core versions.
/// </summary>
public static class CompatibilityChecker
{
    /// <summary>
    /// Überprüft die Kompatibilität der übergebenen Nachricht.<br/>
    /// <u><b>Check compatibility:</b></u> Checks the compatibility of the provided message.
    /// </summary>
    /// <param name="model">
    /// Die zu prüfende Nachricht.<br/>
    /// <u><b>Message:</b></u> The message to check.
    /// </param>
    /// <returns>
    /// Das Ergebnis der Kompatibilitätsprüfung.<br/>
    /// <u><b>Result:</b></u> The result of the compatibility check.
    /// </returns>
    public static CompatibilityResult Check(UebermittlungSchriftgutobjekteNachricht model)
    {
        // Start with all known versions (excluding None)
        var xVersions = Enum.GetValues<XJustizVersion>()
             .Where(v => v != XJustizVersion.None)
             .ToList();

        var xCoreVersions = Enum.GetValues<XJustizCoreVersion>()
             .Where(v => typeof(XJustizCoreVersion).GetField(v.ToString())?.GetCustomAttribute<ForTestingOnlyAttribute>() == null)
             .ToList();

        CheckObject(model, xVersions, xCoreVersions);

        return new CompatibilityResult
        {
            CompatibleXJustizVersions = xVersions,
            CompatibleXJustizCoreVersions = xCoreVersions,
        };
    }

    private static void CheckObject(object? obj, List<XJustizVersion> xVersions, List<XJustizCoreVersion> xCoreVersions)
    {
        if (obj == null)
        {
            return;
        }

        if (xVersions.Count == 0 && xCoreVersions.Count == 0)
        {
            return;
        }

        Type type = obj.GetType();

        // Check Class Attributes (only for our models)
        if (IsXJustizModel(type))
        {
            ApplyAttributes(type.GetCustomAttributes(), xVersions, xCoreVersions);
        }

        CheckProperties(obj, xVersions, xCoreVersions);
    }

    private static void CheckProperties(object obj, List<XJustizVersion> xVersions, List<XJustizCoreVersion> xCoreVersions)
    {
        foreach (var prop in obj.GetType().GetProperties(BindingFlags.Public | BindingFlags.Instance))
        {
            // Skip indexers
            if (prop.GetIndexParameters().Length > 0)
            {
                continue;
            }

            if (TryGetPropertyValue(prop, obj, out var val) && val != null)
            {
                ApplyAttributes(prop.GetCustomAttributes(), xVersions, xCoreVersions);
                CheckValueRecursively(val, xVersions, xCoreVersions);
            }

            // Optimization: Stop if no versions left
            if (xVersions.Count == 0 && xCoreVersions.Count == 0)
            {
                return;
            }
        }
    }

    private static bool TryGetPropertyValue(PropertyInfo prop, object obj, out object? val)
    {
        try
        {
            val = prop.GetValue(obj);
            return true;
        }
        catch
        {
            val = null;
            return false;
        }
    }

    private static void CheckValueRecursively(object val, List<XJustizVersion> xVersions, List<XJustizCoreVersion> xCoreVersions)
    {
        if (val is string)
        {
            // String is IEnumerable but we treat it as primitive
            return;
        }

        if (val is IEnumerable enumerable)
        {
            foreach (var item in enumerable)
            {
                CheckObject(item, xVersions, xCoreVersions);
            }
        }
        else if (IsXJustizModel(val.GetType()))
        {
            CheckObject(val, xVersions, xCoreVersions);
        }
    }

    private static void ApplyAttributes(IEnumerable<Attribute> attributes, List<XJustizVersion> xVersions, List<XJustizCoreVersion> xCoreVersions)
    {
        foreach (var attr in attributes)
        {
            if (attr is XJustizAvailabilityAttribute xAttr)
            {
                // Remove versions that are NOT available
                // ToList() used to allow modification during iteration if we used remove (but RemoveAll handles list fine)
                xVersions.RemoveAll(v => !xAttr.IsAvailableIn(v));
            }
            else if (attr is XJustizCoreAvailabilityAttribute cAttr)
            {
                xCoreVersions.RemoveAll(v => !cAttr.IsAvailableIn(v));
            }
        }
    }

    private static bool IsXJustizModel(Type type)
    {
        // Check if the type is defined in the xjustiz assembly/namespace
        return type.Namespace?.StartsWith("xjustiz", StringComparison.OrdinalIgnoreCase) == true;
    }

    /// <summary>
    /// Das Ergebnis der Kompatibilitätsprüfung.<br/>
    /// <u><b>Compatibility Result:</b></u> The result of the compatibility check.
    /// </summary>
    public class CompatibilityResult
    {
        /// <summary>
        /// Eine Liste der kompatiblen X.Justiz-Versionen.<br/>
        /// <u><b>Compatible X.Justiz versions:</b></u> A list of compatible X.Justiz versions.
        /// </summary>
        public List<XJustizVersion> CompatibleXJustizVersions { get; set; } = [];

        /// <summary>
        /// Eine Liste der kompatiblen X.Justiz-Core-Versionen.<br/>
        /// <u><b>Compatible X.Justiz Core versions:</b></u> A list of compatible X.Justiz Core versions.
        /// </summary>
        public List<XJustizCoreVersion> CompatibleXJustizCoreVersions { get; set; } = [];
    }
}

File: java/src/main/java/de/xjustiz/core/util/versioning/CompatibilityChecker.java
package de.xjustiz.core.util.versioning;

import de.xjustiz.core.models.UebermittlungSchriftgutobjekteNachricht;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Überprüft die Kompatibilität von X.Justiz-Nachrichten mit verschiedenen
 * X.Justiz- und X.Justiz Core Versionen.<br/>
 * <u><b>Compatibility Checker:</b></u> Checks the compatibility of X.Justiz
 * messages with different X.Justiz and X.Justiz Core versions.
 */
public final class CompatibilityChecker {

    private CompatibilityChecker() {
        // Utility class
    }

    /**
     * Überprüft die Kompatibilität der übergebenen Nachricht.<br/>
     * <u><b>Check compatibility:</b></u> Checks the compatibility of the provided
     * message.
     * 
     * @param model Die zu prüfende Nachricht. / The message to check.
     * @return Das Ergebnis der Kompatibilitätsprüfung. / The result of the
     *         compatibility check.
     */
    public static CompatibilityResult check(UebermittlungSchriftgutobjekteNachricht model) {
        // Start with all known versions (excluding NONE for XJustiz and testing
        // versions for Core)
        List<XJustizVersion> xVersions = Arrays.stream(XJustizVersion.values())
                .filter(v -> v != XJustizVersion.NONE)
                .collect(Collectors.toCollection(ArrayList::new));

        List<XJustizCoreVersion> xCoreVersions = Arrays.stream(XJustizCoreVersion.values())
                .filter(v -> !v.isForTestingOnly())
                .collect(Collectors.toCollection(ArrayList::new));

        checkObject(model, xVersions, xCoreVersions);

        return new CompatibilityResult(xVersions, xCoreVersions);
    }

    private static void checkObject(Object obj, List<XJustizVersion> xVersions,
            List<XJustizCoreVersion> xCoreVersions) {
        if (obj == null) {
            return;
        }

        if (xVersions.isEmpty() && xCoreVersions.isEmpty()) {
            return;
        }

        Class<?> type = obj.getClass();

        // Check Class Annotations (only for our models)
        if (isXJustizModel(type)) {
            applyAnnotations(type, xVersions, xCoreVersions);
        }

        checkFields(obj, xVersions, xCoreVersions);
    }

    private static void checkFields(Object obj, List<XJustizVersion> xVersions,
            List<XJustizCoreVersion> xCoreVersions) {
        Class<?> type = obj.getClass();

        // Check all declared fields and inherited fields
        List<Field> allFields = getAllFields(type);

        for (Field field : allFields) {
            field.setAccessible(true);
            Object value = getFieldValue(field, obj);

            if (value != null) {
                applyAnnotations(field, xVersions, xCoreVersions);
                checkValueRecursively(value, xVersions, xCoreVersions);
            }

            // Optimization: Stop if no versions left
            if (xVersions.isEmpty() && xCoreVersions.isEmpty()) {
                return;
            }
        }
    }

    private static List<Field> getAllFields(Class<?> type) {
        List<Field> fields = new ArrayList<>();
        Class<?> current = type;
        // Only traverse our model classes, not JDK classes
        while (current != null && current != Object.class && isXJustizModel(current)) {
            fields.addAll(Arrays.asList(current.getDeclaredFields()));
            current = current.getSuperclass();
        }
        return fields;
    }

    private static Object getFieldValue(Field field, Object obj) {
        try {
            // Skip synthetic and special fields
            if (field.isSynthetic()) {
                return null;
            }
            field.setAccessible(true);
            return field.get(obj);
        } catch (Exception e) {
            // Reflection may fail for various reasons, just skip
            return null;
        }
    }

    private static void checkValueRecursively(Object val, List<XJustizVersion> xVersions,
            List<XJustizCoreVersion> xCoreVersions) {
        if (val instanceof String) {
            // String is Iterable but we treat it as primitive
            return;
        }

        if (val instanceof Collection<?> collection) {
            for (Object item : collection) {
                checkObject(item, xVersions, xCoreVersions);
            }
        } else if (val.getClass().isArray()) {
            Object[] array = (Object[]) val;
            for (Object item : array) {
                checkObject(item, xVersions, xCoreVersions);
            }
        } else if (isXJustizModel(val.getClass())) {
            checkObject(val, xVersions, xCoreVersions);
        }
    }

    private static void applyAnnotations(java.lang.reflect.AnnotatedElement element,
            List<XJustizVersion> xVersions, List<XJustizCoreVersion> xCoreVersions) {

        XJustizAvailability xAvail = element.getAnnotation(XJustizAvailability.class);
        if (xAvail != null) {
            xVersions.removeIf(v -> !VersionAvailabilityChecker.isAvailable(
                    v.getValue(),
                    xAvail.introduced().getValue(),
                    xAvail.removed() == XJustizVersion.NONE ? 0 : xAvail.removed().getValue()));
        }

        XJustizCoreAvailability xcAvail = element.getAnnotation(XJustizCoreAvailability.class);
        if (xcAvail != null) {
            int removedValue = xcAvail.isRemoved() ? xcAvail.removed().getValue() : 0;
            xCoreVersions.removeIf(v -> !VersionAvailabilityChecker.isAvailable(
                    v.getValue(),
                    xcAvail.introduced().getValue(),
                    removedValue));
        }
    }

    private static boolean isXJustizModel(Class<?> type) {
        // Check if the type is defined in the xjustiz package
        String packageName = type.getPackageName();
        return packageName != null && packageName.startsWith("de.xjustiz");
    }
}
